ui/uihandler.go

```diff
@@ -0,0 +1,234 @@
+		if !ok {
+			log.Fatalf("expected http.ResponseWriter to be an http.Flusher")
+		}
+		if DoLoad || DoExit {
+			flusher.Flush()
+		}
+	}
+	if DoLoad {
+		resolution, _ := strconv.ParseFloat(r.FormValue("r"), 64)
+		percList, _ := stats.ParsePercentiles(r.FormValue("p"))
+		qps, _ := strconv.ParseFloat(r.FormValue("qps"), 64)
+		dur, _ := time.ParseDuration(r.FormValue("t"))
+		c, _ := strconv.Atoi(r.FormValue("c"))
+		firstHeader := true
+		for _, header := range r.Form["H"] {
+			if len(header) == 0 {
+				continue
+			}
+			log.LogVf("adding header %v", header)
+			if firstHeader {
+				// If there is at least 1 non empty H passed, reset the header list
+				fhttp.ResetHeaders()
+				firstHeader = false
+			}
+			err := fhttp.AddAndValidateExtraHeader(header)
+			if err != nil {
+				log.Errf("Error adding custom headers: %v", err)
+			}
+		}
+		out := io.Writer(w)
+		if JSONOnly {
+			out = os.Stderr
+		}
+		ro := periodic.RunnerOptions{
+			QPS:         qps,
+			Duration:    dur,
+			Out:         out,
+			NumThreads:  c,
+			Resolution:  resolution,
+			Percentiles: percList,
+		}
+		o := fhttp.HTTPRunnerOptions{
+			RunnerOptions: ro,
+			URL:           url,
+		}
+		res, err := fhttp.RunHTTPTest(&o)
+		if err != nil {
+			w.Write([]byte(fmt.Sprintf("Aborting because %v\n", err))) // nolint: errcheck
+		} else {
+			if JSONOnly {
+				w.Header().Set("Content-Type", "application/json")
+				j, err := json.MarshalIndent(res, "", "  ")
+				if err != nil {
+					log.Fatalf("Unable to json serialize result: %v", err)
+				}
+				_, err = w.Write(j)
+				if err != nil {
+					log.Errf("Unable to write json output for %v: %v", r.RemoteAddr, err)
+				}
+			} else {
+				// nolint: errcheck
+				w.Write([]byte(fmt.Sprintf("All done %d calls %.3f ms avg, %.1f qps\n</pre></body></html>\n",
+					res.Result().DurationHistogram.Count,
+					1000.*res.Result().DurationHistogram.Avg,
+					res.Result().ActualQPS)))
+			}
+		}
+	}
+	if DoExit {
+		syscall.Kill(syscall.Getpid(), syscall.SIGINT) // nolint: errcheck
+	}
+}
+
+// LogoHandler is the handler for the logo
+func LogoHandler(w http.ResponseWriter, r *http.Request) {
+	log.Infof("%v %v %v %v", r.Method, r.URL, r.Proto, r.RemoteAddr)
+	w.Header().Set("Content-Type", "image/svg+xml")
+	// nolint: errcheck, lll
+	w.Write([]byte(`<svg viewBox="0 0 424 650" xmlns="http://www.w3.org/2000/svg"><g fill="#fff"><path d="M422 561l-292 79-118-79 411 0Zm-282-350v280l-138 45 138-325ZM173 11l0 480 250 47-250-527Z"/></g></svg>`))
+}
+
+// Serve starts the fhttp.Serve() plus the UI server on the given port
+// and paths (empty disables the feature).
+func Serve(port int, debugpath string, uiPath string) {
+	debugPath = debugpath
+	startTime = time.Now()
+	httpPort = port
+	if uiPath != "" {
+		http.HandleFunc(uiPath, Handler)
+		fmt.Printf("UI starting - visit:\nhttp://localhost:%d%s\n", port, uiPath)
+		logoPath = uiPath + "/logo.svg"
+		http.HandleFunc(logoPath, LogoHandler)
+	}
+	fhttp.Serve(port, debugpath)
```

Comment:
Do you think that using the same server / port for both the echo server and UI might affect the 'purity' of the stat results? E.g. if in the future you will show some real-time dynamic graph that fetches it from the server, won't it affect server response time to echo requests?

Reply:
that’s why there is a sub path for the UI, with the default config:

/debug is a fast debug
/fortio is this UI
everything else is echo server including / and / echo etc